define type mapping between ROS IDL and C++ / Python

dirk-thomas opened the pull request:
Since this is a critical piece I would like to collect multiple +1's for this before merging it. Please review carefully.

dirk-thomas committed:
add type mapping between ROS IDL and C++ to article

dirk-thomas committed:
add article about generated Python interfaces

esteve commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,108 @@
+
+### "Constants"
+
+A constant is are defined as class variables with uppercase name.
```
s/is are/is/

esteve replied:
Or "Constants are defined as..."

dirk-thomas replied:
FIxed in 8b3f8cfa6803b9cb6b7926c40bb96d86d39ddf2f

esteve commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,108 @@
+
+### "Constants"
+
+A constant is defined as a class variables with uppercase name.
```
A constant is defined as a class variable with an uppercase name.

dirk-thomas replied:
969f528c34def3565f3da7f450472c22ab01decb

esteve commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,108 @@
+
+The *__init__* function initializes all members with their default value.
+
+The class constructor supports keyword arguments for all members.
```
keyword arguments only or in addition to positional arguments?

esteve replied:
I'd prefer only keyword arguments, but it should be noted in the document.

dirk-thomas replied:
969f528c34def3565f3da7f450472c22ab01decb

esteve commented on articles/112_generated_interfaces_cpp.md:
```diff
@@ -50,6 +50,40 @@ For each message two files are being generated:
 This allow to add addition files beside the one with the suffix `__struct` to provide additional functionality.
```
"This allows to add additional files besides the one with the suffix `__struct` to provide additional functionality."

Though there are too many add and additional in this sentence for my taste.

dirk-thomas replied:
Update in fc387348fe8f233bbfb4f1cb77de4aa075e78df4 - even if the line was not even part of this PR.

esteve commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+
+## Messages
+
+For a message a Python `class` with the same name is generated in the file `<my_message_name>.py`.
```
Should this follow the same pattern as ROS1, where messages are generated in `_<my_message_name>.py` and there's an `__init__.py` file that imports all of them? For example:

``` python
from package.msg._Structure import Structure
```

and in `_Structure.py`:

``` python
class Structure:
    pass
```

esteve replied:
Also, is it worth mentioning that, since we're supporting Python 3 only, we can omit `(object)` in the class definition?

dirk-thomas replied:
We could do that.

dirk-thomas replied:
I think the reference to `object` is way to technical for this article. It is at best an implementation detail.

esteve replied:
Yeah, I agree. Do you think we should mention that these types are Python 3 specific? Or is that already implied?

dirk-thomas replied:
That is already mentioned in another design article if I recall correctly.

dirk-thomas replied:
I described the underscore / exporting in 7cd87daf60f5abc72d1a0fe68331bb4252577ebd

esteve commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+| uint32   | builtins.int                 |
+| int64    | builtins.int                 |
+| uint64   | builtins.int                 |
+| string   | builtins.str                 |
```
Could you add a `TODO` or a note about encoding and that Python 3 only uses unicode for strings?

dirk-thomas replied:
The TODO is already in the ROSIDL article. While it affects all languages I am not sure we need to duplicate the todos.

wjwwood replied:
Until the TODO in the rosidl article is resolved we either have to use unicode or not, so I think there should be a TODO here saying what we're going to do for now and mention that it may change or be clarified once that TODO is resolved. So I think a TODO here is a valuable, but temporary, clarification which isn't redundant.

wjwwood replied:
Also, I think [`bytearray`](https://docs.python.org/3.1/library/functions.html#bytearray) might be a better match for string in rosidl. This forces the user to create a bytearray by providing their string and the expected encoding. The result can be stored in a `std::string` and `char *` when deserialized on the other side. I think mapping a Python3 string into a rosidl string (and therefore a IDL string) and hoping it doesn't have unicode in it is a bad idea. Put another way I think the problem should come up in Python code on assignment rather than on publish.

wjwwood replied:
Also, since we're using Python 3.5+ (something that isn't mentioned in the article yet and I think should since it affects the mapping between types) we can use VarType and properties to help enforce this: https://docs.python.org/3/library/typing.html#typing.TypeVar

For example if we add `wstring` to rosidl then we can have that map to `TypeVar('wstring', bytearray, str)`, since `wstring` is a superset of `string`.

This might not end up working, but these are just some thoughts I had.

dirk-thomas replied:
I agree that any encoding check should happen in pure Python code. Simply because the goal is that the Python data structure are usable (and consistent) when being used independently.

We are not using Python 3.5 right now since the current version on Trusty is 3.4. But we should keep this in mind when updating the dependencies with Xenial.

wjwwood commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+
+### "Constants"
+
+A constant is defined as a class variable with uppercase name.
```
Are they read-only?

dirk-thomas replied:
Yes, good point: 5f23395054c82b14c433a42ea59e84afefac56f6

wjwwood commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+The *__init__* function initializes all members with their default value.
+
+The class constructor supports only keyword arguments for all members.
+The short reason for this is that if code would rely on positional arguments to construct data objects changing a message definition would break existing code in various ways.
```
Instead of "various ways" I'd say "subtle ways" or "easily overlooked ways".

dirk-thomas replied:
Done: 8af72032a1fbb5db90e79f3010137ed9a6fa686

wjwwood commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+
+### Request and response messages
+
+For the request as well as the response part messages are being generated which have the same name as the service with a suffix being either `_Request` or `_Response`.
```
This sentence was awkward for me to read, how about: "The request and response components for a Service are messages whose names are made up of the Service's name and either a `_Request` or `_Response` suffix respectively."

dirk-thomas replied:
Rephrased in 696173c9e6eb5f9f896d3138394fc3e29117c828

wjwwood commented on articles/114_generated_interfaces_python.md:
```diff
@@ -0,0 +1,110 @@
+
+## Namespacing
+
+All code of a ROS package should be defined in a namespace named after the package.
```
Should we use the term `namespace` here or the Pythonic term `package`? Perhaps we should continue to use `namespace`, but clarify that what we mean is within in a Python package there is a `msg` and/or `srv` sub-package in which there is a Python module for each message or service which contains a Python class.

dirk-thomas replied:
I rephrased it: 082b3bdb8331a23319364425fda1e8f84bfc84f6

dirk-thomas replied:
Wrong file: a7b3fe8d4c4248fbc461cd1c0adcb769f337c7cb

ernestmc commented on articles/112_generated_interfaces_cpp.md:
```diff
@@ -47,9 +47,43 @@ For each message two files are being generated:
+| ----------------------- | ---------------- |
+| static array            | std::array<T, N> |
+| unbounded dynamic array | std::vector<T>   |
+| bounded dynamic array   | std::vector<T>   |
```
Currently bounded arrays are mapped as std::array<T, N>

dirk-thomas replied:
Indeed, but I think we have to revisit this. An array always has N elements but we need to be able to store explicitly only M (<N) elements (and are able to get the count M).

I added a TODO for that: 84131cfa908ab465b7ae541612700183014f3843

dirk-thomas commented:
I have addressed all the feedback. I will rewrite this branch since the number of commits affecting two different articles becomes a bit messy. If you already have a fork of this make sure to pull the latest version.

wjwwood commented:
I still think a TODO about unicode (and what we're doing in the mean time) is a good idea.

Otherwise, lgtm.

dirk-thomas committed:
add TODO about encoding of strings being unspecified

dirk-thomas commented:
I added a TODO for this in the language agnostic article: d67f9dfd952f33e962815061b3ebb69f4cca8c1e

esteve commented:
+1

esteve commented:
Though I agree with @wjwwood that a TODO here about the Python-specific issue of `str` vs `unicode` types is valuable.

esteve commented:
@dirk-thomas Whoa, that was fast :-)
